### PR TITLE
[RHELC-1102] feat: Initial testing of el9 conversions

### LINKED ISSUE #867: Support for EL9 based clones to convert to RHEL9
Hello, is there also a plan to be able to migrate from EL9 based distributions such as (AlmaLinux, Rocky, Oracle) to RHEL9?

Thank you.

### PR DESCRIPTION
**Pull request by SpyTec**
Primarily to see what needs to be changed, if any, to get el9 conversions to work.

Fixes #867

Todo

- [ ] From @abadger: The CentOS 9 output looks like the NEVR format used in an internal dict doesn't match with the output that we're getting from something else.  I wonder if the format from dnf changed in rhel9

Jira Issues: [RHELC-1102](https://issues.redhat.com/browse/RHELC-1102)

Checklist

- [ ] PR has been tested manually in a VM (either author or reviewer)
- [x] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [x] GitHub label has been added to help with Release notes 
- [ ] PR title explains the change from the user's point of view
- [ ] Code and tests are documented properly
- [ ] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

### CODE REVIEW

**Comment by vrutkovs** [comment]
This doesn't seem to work on AlmaLinux 9:
```
[07/28/2023 10:23:01] TASK - [Prepare: Check kernel compatibility with RHEL] ********************
[07/28/2023 10:23:01] DEBUG - Unexpected OS major version. Expected: dict_keys([7, 8])
[07/28/2023 10:23:01] DEBUG - Calling command 'rpm -qf --qf %{VERSION}&%{RELEASE}&%{ARCH}&%{NAME} /boot/vmlinuz-5.14.0-284.18.1.el9_2.x86_64'
[07/28/2023 10:23:01] DEBUG - Booted kernel package name: kernel-core
WARNING - Custom kernel detected. The booted kernel needs to be signed by AlmaLinux.
ERROR - (ERROR) RHEL_COMPATIBLE_KERNEL.BOOTED_KERNEL_INCOMPATIBLE: The booted kernel version is incompatible with the standard RHEL kernel. To proceed with the conversion, boot into a kernel that is available in the AlmaLinux 9 base repository by executing the following steps:

1. Ensure that the AlmaLinux 9 base repository is enabled
2. Run: yum install kernel
3. (optional) Run: grubby --set-default /boot/vmlinuz-`rpm -q --qf "%{BUILDTIME}\t%{EVR}.%{ARCH}\n" kernel | sort -nr | head -1 | cut -f2`
4. Reboot the machine and if step 3 was not applied choose the kernel installed in step 2 manually
```
and then
```
[07/28/2023 10:23:01] TASK - [Convert: List third-party packages] *******************************
WARNING - Only packages signed by AlmaLinux are to be replaced. Red Hat support won't be provided for the following third party packages:

ERROR - (ERROR) LIST_THIRD_PARTY_PACKAGES.UNEXPECTED_ERROR: Unhandled exception was caught: '0:acl-2.3.1-3.el9.x86_64'
Please file a bug at https://issues.redhat.com/ to have this fixed or a specific error message added.
Traceback: Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/convert2rhel/actions/__init__.py", line 393, in run
    action.run()
  File "/usr/lib/python3.9/site-packages/convert2rhel/actions/pre_ponr_changes/handle_packages.py", line 45, in run
    pkghandler.print_pkg_info(third_party_pkgs)
  File "/usr/lib/python3.9/site-packages/convert2rhel/utils.py", line 245, in wrapper
    raise process.exception
KeyError: '0:acl-2.3.1-3.el9.x86_64'
```

Seems it needs to account for kernel to come from `kernel-core` package - and AlmaLinux GPG key is mismatching(?)

**Comment by SpyTec** [comment]
@vrutkovs it seems like it is related to a bug with a mismatch with package info. @r0x0d brought this up to me:

> Probably must be this part of the code: https://github.com/oamg/convert2rhel/blob/8d8a9c1587ef01acce1279a5865624905103e33d/convert2rhel/pkghandler.py#L408-L412
> 
> I think that what is in package_info is different from what is in the packages_with_repos, if so, should be easy to fix.
> 
> This is the condition we have to retrieve information from repoquery: https://github.com/oamg/convert2rhel/blob/8d8a9c1587ef01acce1279a5865624905103e33d/convert2rhel/pkghandler.py#L442-L451

**Comment by SpyTec** [comment]
Oh and as for the first error I see

> [07/28/2023 10:23:01] DEBUG - Unexpected OS major version. Expected: dict_keys([7, 8])

Which seems to be me forgetting to add the RHEL kernel version

https://github.com/SpyTec/convert2rhel/blob/d33de304bee4a5e3653a44e5d6f6a5f2b8206c70/convert2rhel/actions/system_checks/rhel_compatible_kernel.py#L29-L32

**Review by vrutkovs — CHANGES_REQUESTED** [review]

**Comment by kdknigga** [comment]
https://github.com/oamg/convert2rhel/blob/6476ef8260919324c81fea3e56b6a9bbe17ba7b7/convert2rhel/subscription.py#L906

Another thing that needs updating.

**Comment by kdknigga** [comment]
https://github.com/oamg/convert2rhel/blob/6476ef8260919324c81fea3e56b6a9bbe17ba7b7/convert2rhel/subscription.py#L50

We'll need UBI_9 repo configuration.

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/data/9/x86_64/rhel-certs/486.pem [hunk 1]
```diff
@@ -0,0 +1,35 @@
+-----BEGIN CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (bocekm) [reply convert2rhel/data/9/x86_64/rhel-certs/486.pem:1]: It should be taken from the redhat-release package instead.

File: convert2rhel/data/9/x86_64/configs/almalinux-9-x86_64.cfg [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# The GPG key is available at https://repo.almalinux.org/almalinux/RPM-GPG-KEY-AlmaLinux
+# Delimited by whitespace(s).
+gpg_fingerprints = 488fcf7c3abb34f8
+
+# List of packages to be removed before the system conversion starts.
+# Delimited by any whitespace(s).
+excluded_pkgs =
+  almalinux-logos*
+  almalinux-indexhtml
+  almalinux-gpg-keys
```
[the remark this answers is not in the document]
Comment (bocekm) [reply convert2rhel/data/9/x86_64/configs/almalinux-9-x86_64.cfg:13]: Thanks, I'll check that when converting Alma 9 how it behaves or where it fails exactly.

**Comment by bocekm** [comment]
Thanks for the reviews, @vrutkovs and @kdknigga. I've done the changes you proposed and will look further at what's failing next.

File: convert2rhel/subscription.py [hunk 3]
```diff
@@ -901,9 +916,8 @@ def download_rhsm_pkgs():
 
     if system_info.version.major == 7:
         pkgs_to_download += ["subscription-manager-rhsm", "python-syspurpose"]
-        _download_rhsm_pkgs(pkgs_to_download, _UBI_7_REPO_PATH, _UBI_7_REPO_CONTENT)
 
-    elif system_info.version.major == 8:
+    elif system_info.version.major >= 8:
         pkgs_to_download += [
             "python3-subscription-manager-rhsm",
             "dnf-plugin-subscription-manager",
```
Comment (kdknigga) [inline convert2rhel/subscription.py:923]: The package in el9 is named `libdnf-plugin-subscription-manager`.

```
$ sudo dnf --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 repolist                              
repo id                                   repo name
ubi-9-baseos-convert2rhel                 Red Hat Universal Base Image 9 - BaseOS added by Convert2RHEL

$ sudo dnf --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 search dnf-plugin-subscription-manager
Last metadata expiration check: 0:06:29 ago on Fri 11 Aug 2023 09:04:18 AM CDT.
============== Name Matched: dnf-plugin-subscription-manager ==============
libdnf-plugin-subscription-manager.x86_64 : Subscription Manager plugin for libdnf
``

Comment (kdknigga) [reply convert2rhel/subscription.py:923]: Also, python3-syspurpose doesn't exist in el9.

Comment (kdknigga) [reply convert2rhel/subscription.py:923]: Also, `subscription-manager-rhsm-certificates` matches two packages:

```
$ sudo rpm -qip *.rpm
Name        : subscription-manager-rhsm-certificates
Version     : 1.29.26.1
Release     : 1.el9_0
Architecture: x86_64
Install Date: (not installed)
Group       : Development/Libraries
Size        : 9716
License     : GPLv2
Signature   : RSA/SHA256, Tue 19 Jul 2022 08:59:04 AM CDT, Key ID 199e2f91fd431d51
Source RPM  : subscription-manager-1.29.26.1-1.el9_0.src.rpm
Build Date  : Mon 18 Jul 2022 02:21:29 PM CDT
Build Host  : x86-64-02.build.eng.rdu2.redhat.com
Packager    : Red Hat, Inc. <http://bugzilla.redhat.com/bugzilla>
Vendor      : Red Hat, Inc.
URL         : http://www.candlepinproject.org/
Summary     : Certificates required to communicate with a Red Hat Unified Entitlement Platform
Description :
This package contains certificates required for communicating with the REST interface
of a Red Hat Unified Entitlement Platform, used for the management of system entitlements
and to receive access to content.
Name        : subscription-manager-rhsm-certificates
Version     : 20220623
Release     : 1.el9
Architecture: noarch
Install Date: (not installed)
Group       : Unspecified
Size        : 27808
License     : GPLv2
Signature   : RSA/SHA256, Tue 28 Jun 2022 04:00:39 AM CDT, Key ID 199e2f91fd431d51
Source RPM  : subscription-manager-rhsm-certificates-20220623-1.el9.src.rpm
Build Date  : Thu 23 Jun 2022 08:07:45 AM CDT
Build Host  : ppc-002.build.eng.bos.redhat.com
Packager    : Red Hat, Inc. <http://bugzilla.redhat.com/bugzilla>
Vendor      : Red Hat, Inc.
URL         : https://www.candlepinproject.org/
Summary     : Certificates required to communicate with a Red Hat Unified Entitlement Platform
Description :
This package contains certificates required for communicating with the REST interface
of a Red Hat Unified Entitlement Platform, used for the management of system entitlements
and to receive access to content.
```

Comparing to a machine built originally with RHEL9 that I have, it looks like `subscription-manager-rhsm-certificates-20220623` might be the correct choice.  Forcing that worked.

Comment (bocekm) [reply convert2rhel/subscription.py:923]: Thanks, I was looking for all RHEL 8-related things and changing them without looking into details. Let me change this one now.

Comment (kdknigga) [reply convert2rhel/subscription.py:923]: Once I hacked my way through these three things, I got a successful conversion from Alma 9 to RHEL 9.  So, it's close!

Comment (bocekm) [reply convert2rhel/subscription.py]: The version of `subscription-manager-rhsm-certificates-20220623` is considered newer than the `subscription-manager-rhsm-certificates-1.29.26*` packages by dnf.

We're using `yumdownloader` which uses dnf under the hood, so it's going to pick up the right one when downloading it - the `20220623`.

Comment (kdknigga) [reply convert2rhel/subscription.py]: It does not pick up the right one, it gets both.

```
[2023-08-11T09:56:26-0500] TASK - [Convert: Subscription Manager - Download packages] ****************                                                                                                                                        
[2023-08-11T09:56:26-0500] DEBUG - Contents of /usr/share/convert2rhel/subscription-manager directory before RHEL rhsm packages download:                                                                                                     
                                                                                                                                                                                                                                              
[2023-08-11T09:56:26-0500] DEBUG - Downloading the subscription-manager package.                                                                                                                                                              
[2023-08-11T09:56:26-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 subscriptio
n-manager'                                                                                                                                                                                                                                    
Successfully downloaded the subscription-manager package.                                                                                                                                                                                     
[2023-08-11T09:56:27-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm                                                                             
[2023-08-11T09:56:27-0500] DEBUG - Downloading the subscription-manager-rhsm-certificates package.                                                                                                                                            
[2023-08-11T09:56:27-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 subscriptio
n-manager-rhsm-certificates'                                                                                                                                                                                                                  
Successfully downloaded the subscription-manager-rhsm-certificates package.                                                                                                                                                                   
[2023-08-11T09:56:30-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64.rpm                                                           
[2023-08-11T09:56:30-0500] DEBUG - Downloading the python3-subscription-manager-rhsm package.                                                                                                                                                 
[2023-08-11T09:56:30-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 python3-sub
scription-manager-rhsm'                                                                                                                                                                                                                       
Successfully downloaded the python3-subscription-manager-rhsm package.                                                                                                                                                                        
[2023-08-11T09:56:30-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/python3-subscription-manager-rhsm-1.29.33.1-1.el9_2.x86_64.rpm                                                                
[2023-08-11T09:56:30-0500] DEBUG - Downloading the libdnf-plugin-subscription-manager package.                                                                                                                                                
[2023-08-11T09:56:30-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 libdnf-plug
in-subscription-manager'                                                                                                                                                                                                                      
Successfully downloaded the libdnf-plugin-subscription-manager package.                                                                                                                                                                       
[2023-08-11T09:56:31-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/libdnf-plugin-subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm                                                               
[2023-08-11T09:56:31-0500] DEBUG - Downloading the python3-cloud-what package.                                                                                                                                                                
[2023-08-11T09:56:31-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 python3-clo
ud-what'                                                                                                                                                                                                                                      
Successfully downloaded the python3-cloud-what package.                                                                                                                                                                                       
[2023-08-11T09:56:31-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/python3-cloud-what-1.29.33.1-1.el9_2.x86_64.rpm                                                                               
[2023-08-11T09:56:31-0500] DEBUG - Downloading the json-c.x86_64 package.                                                                                                                                                                     
[2023-08-11T09:56:31-0500] DEBUG - Calling command 'yumdownloader -v --destdir=/usr/share/convert2rhel/subscription-manager --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 json-c.x86_
64'                                                                                                                                                                                                                                           
Successfully downloaded the json-c.x86_64 package.                                                                                                                                                                                            
[2023-08-11T09:56:32-0500] DEBUG - Path of the downloaded package: /usr/share/convert2rhel/subscription-manager/json-c-0.14-11.el9.x86_64.rpm                                                                                                 
[2023-08-11T09:56:32-0500] DEBUG - Contents of /usr/share/convert2rhel/subscription-manager directory after RHEL rhsm packages download:                                                                                                      
subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm                                                                                                                                                                                             
subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64.rpm                                                                                                                                                                           
subscription-manager-rhsm-certificates-20220623-1.el9.noarch.rpm                                                                                                                                                                              
python3-subscription-manager-rhsm-1.29.33.1-1.el9_2.x86_64.rpm                                                                                                                                                                                
libdnf-plugin-subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm                                                                                                                                                                               
python3-cloud-what-1.29.33.1-1.el9_2.x86_64.rpm
json-c-0.14-11.el9.x86_64.rpm

[2023-08-11T09:56:32-0500] TASK - [Convert: Subscription Manager - Replace] **************************
Unregistering the system.
[2023-08-11T09:56:32-0500] DEBUG - Calling command 'rpm --quiet -q subscription-manager'
The subscription-manager package is not installed.
Removing installed subscription-manager/katello-ca-consumer packages.
No packages related to subscription-manager installed.
Checking for subscription-manager RPMs.
Installing subscription-manager RPMs.
[2023-08-11T09:56:32-0500] DEBUG - Calling command 'yum install -y --setopt=module_platform_id=platform:el9 /usr/share/convert2rhel/subscription-manager/subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm /usr/share/convert2rhel/subscriptio
n-manager/subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64.rpm /usr/share/convert2rhel/subscription-manager/subscription-manager-rhsm-certificates-20220623-1.el9.noarch.rpm /usr/share/convert2rhel/subscription-manager/pytho
n3-subscription-manager-rhsm-1.29.33.1-1.el9_2.x86_64.rpm /usr/share/convert2rhel/subscription-manager/libdnf-plugin-subscription-manager-1.29.33.1-1.el9_2.x86_64.rpm /usr/share/convert2rhel/subscription-manager/python3-cloud-what-1.29.33
.1-1.el9_2.x86_64.rpm /usr/share/convert2rhel/subscription-manager/json-c-0.14-11.el9.x86_64.rpm'
Last metadata expiration check: 0:00:07 ago on Fri Aug 11 09:56:25 2023.
Package json-c-0.14-11.el9.x86_64 is already installed.
Error: 
 Problem: cannot install both subscription-manager-rhsm-certificates-20220623-1.el9.noarch and subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64
  - conflicting requests
(try to add '--allowerasing' to command line to replace conflicting packages or '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
CRITICAL - Failed to install subscription-manager packages. See the above yum output for details.
ERROR - (ERROR) PRE_SUBSCRIPTION.UNKNOWN_ERROR: Failed to install subscription-manager packages. See the above yum output for details.
```

```
$ sudo yumdownloader -v --destdir=/tmp/rpms --setopt=reposdir=/var/lib/convert2rhel/rhsm --releasever=9.2 --setopt=module_platform_id=platform:el9 subscription-manager-rhsm-certificates           
Loaded plugins: builddep, changelog, config-manager, copr, debug, debuginfo-install, download, generate_completion_cache, groups-manager, needs-restarting, playground, repoclosure, repodiff, repograph, repomanage, reposync, system-upgrade
DNF version: 4.14.0                                                                                                                                                                                                                           cachedir: /var/cache/dnf                                                                                               
User-Agent: constructed: 'libdnf (AlmaLinux 9.2; generic; Linux.x86_64)'
repo: using cache for: ubi-9-baseos-convert2rhel                                                                       
ubi-9-baseos-convert2rhel: using metadata from Tue 08 Aug 2023 10:35:13 AM CDT.
Last metadata expiration check: 0:05:33 ago on Fri 11 Aug 2023 09:56:26 AM CDT.                                                                                                                                                               
(1/2): subscription-manager-rhsm-certificates-20220623-1.el9.noarch.rpm                                 162 kB/s |  22 kB     00:00     
(2/2): subscription-manager-rhsm-certificates-1.29.26.1-1.el9_0.x86_64.rpm                               118 kB/s |  56 kB     00:00    
```

Comment (bocekm) [reply convert2rhel/subscription.py]: Interesting, I'll ask around.

Comment (bocekm) [reply convert2rhel/subscription.py]: Before I get the answer from the subscription-manager folks, I've specified that it should be the x86_64 one that's to be downloaded. That the arch that UBI 7 and 8 have and it was the easiest to do in the code for now. The only difference is that `20220623` one installs this one additional file: `/usr/share/licenses/subscription-manager-rhsm-certificates/COPYING`.

**Comment by SpyTec** [comment]
@bocekm When you can approve and merge to the feature branch so we can collab better

**Comment by SpyTec** [comment]
Created new PR here with a branch in the convert2rhel repo, should be easier to collaborate now
#867

**Pull request merged**